### PR TITLE
Closes #667 Thin Edge JSON alarm support

### LINKED ISSUE #667: As a device owner I want to push raised alarms (drop1)
As a thin-edge device owner I want to be able to raise alarms and push these to the cloud.

Therefore thin-edge public alarm API (MQTT topic "tedge/alarms") shall be introduced. C8Y Mapper to be extended to support that API with thin-edge JSON (alarm) format, as specified in
    https://github.com/thin-edge/thin-edge.io-specs/blob/82d4dd269887ab69544df616aa9244752319f8cb/src/alarms/alarm-handling.md

AC:
  - Scope of that ticket is realization of alarms support for C8Y. Other clouds (e.g. Azure) are not yet in scope.
  - C8Y mapper shall support thin-edge JSON alarm format (as defined in spec linked above) 
  - C8Y mapper shall avoid duplicates (as defined in spec linked above) 
  - Reliable Data chain shall be provided (as defined in spec linked above) 
  - Both functionalities "Raw payload format" and "Addressing Child-Devices" from Spec above, are not in scope of that drop1 implementation.

### PR DESCRIPTION
**Pull request by albinsuresh**
## Proposed changes

Extend C8Y mapper to support thin edge alarms. Thin Edge alarms are directly mapped to the respective Cumulocity equivalents using SmartREST. Deduplication is not done.

## Types of changes

What types of changes does your code introduce to `thin-edge.io`?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Improvement (general improvements like code refactoring that doesn't explicitly fix a bug or add any new functionality)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Paste Link to the issue
<br/>

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] I have read the [CONTRIBUTING](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTOR-LICENSE-AGREEMENT.md) (in all commits with git commit -s)
- [x] I ran `cargo fmt` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [ ] I used `cargo clippy` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

File: crates/core/tedge_mapper/src/c8y_converter.rs [hunk 1]
```diff
@@ -79,6 +75,51 @@ impl Converter for CumulocityConverter {
...
+    }
+
+    fn try_convert(&mut self, input: &Message) -> Result<Vec<Message>, ConversionError> {
+        let () = self.size_threshold.validate(input.payload_str()?)?;
+        if input.topic.name.starts_with("tedge/measurement") {
+            self.try_convert_measurement(input)
+        } else if input.topic.name.starts_with("tedge/alarms") {
+            self.try_convert_alarm(input)
+        } else {
+            return Ok(vec![]);
```
Comment (rina23q) [inline crates/core/tedge_mapper/src/c8y_converter.rs:119]: If unknown topic, shouldn't we return a `ConversionError::UnsupporedTopic`?

Comment (rina23q) [reply crates/core/tedge_mapper/src/c8y_converter.rs:119]: If you see `fn wrap_error()` in `coverter.rs`, you can see that we convert the Err to error messages to publish on `tedge/errors`. I'm sure that we should return `Err` here.

**Review by rina23q — COMMENTED** [review]
First code review. I'm going to test manually.

**Comment by rina23q** [comment]
I tested. It works great!

File: crates/core/tedge_mapper/src/c8y_converter.rs [hunk 2]
```diff
@@ -79,6 +75,51 @@ impl Converter for CumulocityConverter {
...
+        let topic_split: Vec<&str> = topic_name.split('/').collect();
+        if topic_split.len() == 4 {
+            let alarm_name = topic_split[3];
+            let alarm_severity = topic_split[2];
+
+            let c8y_json_payload = smartrest::from_thin_edge_alarm_json(
+                alarm_name,
+                alarm_severity,
+                input.payload_str()?,
+            )?;
```
Comment (didier-wenzek) [inline crates/core/tedge_mapper/src/c8y_converter.rs:95]: Why not push the topic parsing logic inside the `smartrest::from_thin_edge_alarm_json` method?

File: crates/core/c8y_translator/src/smartrest.rs [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+use serde::{Deserialize, Serialize};
+use serde_json;
+
+use crate::json::CumulocityJsonError;
+
+#[derive(Debug, Serialize, Deserialize)]
+pub struct ThinEdgeJsonAlarm {
+    pub message: Option<String>,
+    pub time: Option<DateTime<FixedOffset>>,
+}
```
Comment (didier-wenzek) [inline crates/core/c8y_translator/src/smartrest.rs:12]: This struct feels incomplete, ignoring the key attributes of an alarm.

Why not something along the following lines to separate the keys from the content of the alarm.
The former will be extract from the topic and the latter deserialized from the payload.

```suggestion
pub struct ThinEdgeJsonAlarm {
    pub name: String,
    pub severity: Severity,
    pub event: ThinEdgeJsonAlarmEvent,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct ThinEdgeJsonAlarmEvent {
    pub message: Option<String>,
    pub time: Option<DateTime<FixedOffset>>,
}
```

File: crates/core/c8y_translator/src/smartrest.rs [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+
+        let smartrest_code = match severity {
+            "critical" => 301,
+            "major" => 302,
+            "minor" => 303,
+            "warning" => 304,
+            invalid => Err(CumulocityJsonError::UnsupportedAlarmSeverity(
+                invalid.into(),
+            ))?,
+        };
```
Comment (didier-wenzek) [inline crates/core/c8y_translator/src/smartrest.rs:34]: This should be moved into a `Severity` enum.

File: crates/core/c8y_translator/src/smartrest.rs (hunk 3 above)
Comment (didier-wenzek) [reply crates/core/c8y_translator/src/smartrest.rs:12]: Oh. I just saw `crates/core/thin_edge_json/src/alarm.rs`.

The structs that are defined there must be used here. The deserialization from thin-edge-json must be move there.
And only the smartrest serialization should stay here.

**Review by didier-wenzek — COMMENTED** [review]
The code can be improved with a better separation between the thin-edge-json and smartrest logic.

File: crates/core/thin_edge_json/src/alarm.rs [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+    deserializer: D,
+) -> Result<Option<OffsetDateTime>, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    let timestamp = String::deserialize(deserializer)?;
+    OffsetDateTime::parse(timestamp.as_str(), &Rfc3339)
+        .map_err(serde::de::Error::custom)
+        .map(|val| Some(val))
+}
```
Comment (rina23q) [inline crates/core/thin_edge_json/src/alarm.rs:41]: I think this function should move to `common/tedge_utils`.

**Review by rina23q — COMMENTED** [review]
The code is very neat, however two corner case issues.

1. If the message contains `,`, the mapper generates a kind of broken SmartREST message.
2. We should address the topic `tedge/alarms/<severity>/` as an error.

**Review by rina23q — APPROVED** [review]
Perfect! Thanks Albin!

**Pull request merged**